comm/comm.1
.\"
.\" Derived from comm(1), Unix 32V:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.\"
.TH COMM 1 "3/16/03" "Heirloom Toolchest" "User Commands"
.SH NAME
comm \- select or reject lines common to two sorted files
.SH SYNOPSIS
\fBcomm\fR [\fB\-123\fR] \fIfile1 file2\fR
.SH DESCRIPTION
.I Comm
reads
.I file1
and
.I file2,
which should be ordered in the current collating sequence,
and produces a three column output: lines only in
.I file1;
lines only in
.I file2;
and lines in both files.
The filename `\-' means the standard input.
.PP
Flags 1, 2, or 3 suppress printing of the corresponding
column.
Thus
.B comm
.B \-12
prints only the lines common to the two files;
.B comm
.B \-23
prints only lines in the first file but not in the second;
.B comm
.B \-123
is a no-op.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_COLLATE
Sets the collation sequence.
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters.
.SH "SEE ALSO"
cmp(1),
diff(1),
uniq(1),
locale(7)


CHANGES
Release 250220
* seq, write: Fix memory fault due to illegal pointer access.
* cpio, pax: Fix manual page formatting.

Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.

comm/comm.c
/*
 * Changes by Gunnar Ritter, Freiburg i. Br., Germany, March 2003.
 */
/*	from Unix 32V /usr/src/cmd/comm.c	*/
/*
 * Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
 *
 * SPDX-Licence-Identifier: Caldera
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)comm.sl	1.7 (gritter) 5/29/05";

#include <stdio.h>
#include <iblok.h>
#include <libgen.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <unistd.h>

#include "mbtowi.h"

static int	one;
static int	two;
static int	three;

static char	*ldr[3];

struct iblok	*ib1;
struct iblok	*ib2;

static int	ccoll;
static int	mb_cur_max;

static char	*progname;

static int		rd(struct iblok *, char **, size_t *);
static void		wr(const char *, int);
static void		copy(struct iblok *, char **, size_t *, int);
static int		compare(const char *, const char *);
static struct iblok	*openfil(const char *);
static void		*srealloc(void *, size_t);

#define	next(wc, s, n)	(*(s) & 0200 ? ((n) = mbtowi(&(wc), (s), mb_cur_max), \
		(n) = ((n) > 0 ? (n) : (n) < 0 ? (wc=WEOF, 1) : 1)) : \
	((wc) = *(s) & 0377, (n) = 1))

static void
usage(void)
{
	fprintf(stderr, "usage: %s [ - [123] ] file1 file2\n", progname);
	exit(2);
}

int
main(int argc, char **argv)
{
	int	i, l;
	char	*cols;
	char	*lb1 = NULL, *lb2 = NULL;
	size_t	sb1 = 0, sb2 = 0;

	ldr[0] = strdup("");
	ldr[1] = strdup("\t");
	ldr[2] = strdup("\t\t");
	l = 1;
	progname = basename(argv[0]);
	cols = setlocale(LC_COLLATE, "");
	ccoll = cols == 0 || strcmp(cols, "C") == 0 ||
		strcmp(cols, "POSIX") == 0;
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	while ((i = getopt(argc, argv, ":123")) != EOF) {
		switch (i) {
		case'1':
			if(!one) {
				one = 1;
				ldr[1][0] = ldr[2][l--] = '\0';
			}
			break;
		case '2':
			if(!two) {
				two = 1;
				ldr[2][l--] = '\0';
			}
			break;
		case '3':
			three = 1;
			break;
		default:
			usage();
		}
	}

	argv += optind, argc -= optind;
	if(argc < 2)
		usage();

	ib1 = openfil(argv[0]);
	ib2 = openfil(argv[1]);


	if(rd(ib1,&lb1,&sb1) < 0) {
		if(rd(ib2,&lb2,&sb2) < 0)	exit(0);
		copy(ib2,&lb2,&sb2,2);
	}
	if(rd(ib2,&lb2,&sb2) < 0)
		copy(ib1,&lb1,&sb1,1);

	while(1) {

		switch(compare(lb1,lb2)) {

			case 0:
				wr(lb1,3);
				if(rd(ib1,&lb1,&sb1) < 0) {
					if(rd(ib2,&lb2,&sb2) < 0)	exit(0);
					copy(ib2,&lb2,&sb2,2);
				}
				if(rd(ib2,&lb2,&sb2) < 0)
					copy(ib1,&lb1,&sb1,1);
				continue;

			case 1:
				wr(lb1,1);
				if(rd(ib1,&lb1,&sb1) < 0)
					copy(ib2,&lb2,&sb2,2);
				continue;

			case 2:
				wr(lb2,2);
				if(rd(ib2,&lb2,&sb2) < 0)
					copy(ib1,&lb1,&sb1,1);
				continue;
		}
	}
}

static int
rd(struct iblok *ip, char **lp, size_t *sz)
{
	size_t	len;

	if ((len = ib_getlin(ip, lp, sz, srealloc)) == 0)
		return -1;
	if ((*lp)[len-1] == '\n')
		(*lp)[len-1] = '\0';
	return 0;
}

static void
wr(const char *str, int n)
{

	switch(n) {

		case 1:
			if(one)	return;
			break;

		case 2:
			if(two)	return;
			break;

		case 3:
			if(three)	return;
	}
	printf("%s%s\n",ldr[n-1],str);
}

static void
copy(struct iblok *ip, char **lp, size_t *sz, int n)
{
	do {
		wr(*lp, n);
	} while (rd(ip, lp, sz) >= 0);

	exit(0);
}

static int
compare(const char *a, const char *b)
{
	const char *ra,*rb;
	int	n;

	if (ccoll) {
		if (mb_cur_max > 1) {
			wint_t	wa, wb;
			int	ma, mb;

			while (next(wa, a, ma), next(wb, b, mb), wa == wb) {
				if (wa == '\0')
					return 0;
				a += ma;
				b += mb;
			}
			if (wa < wb)
				return 1;
			return 2;
		} else {
			ra = --a;
			rb = --b;
			while(*++ra == *++rb)
				if(*ra == '\0')	return(0);
			if(*ra < *rb)	return(1);
			return(2);
		}
	} else {
		n = strcoll(a, b);
		return n ? n > 0 ? 2 : 1 : 0;
	}
}

static struct iblok *
openfil(const char *s)
{
	struct iblok	*ip;

	if (s[0] == '-' && s[1] == '\0')
		ip = ib_alloc(0, 0);
	else if ((ip = ib_open(s, 0)) == NULL) {
		fprintf(stderr,"%s: cannot open %s\n", progname, s);
		exit(1);
	}
	return ip;
}

static void *
srealloc(void *op, size_t sz)
{
	void	*np;

	if ((np = realloc(op, sz)) == NULL) {
		write(2, "no memory\n", 10);
		_exit(077);
	}
	return np;
}
